src/datasource.ts

```diff
@@ -271,8 +271,12 @@ export default class HeroicDatasource {
     options.headers = headers;
     options.url = this.settings.url + path;
     options.inspect = { type: "heroic" };
+    if (options.method === 'POST' && options.data) {
+      options.data = JSON.stringify(options.data);
```

Comment:
what happens with older versions of Grafana? will this get double stringified?

Reply:
> will this get double stringified?

Good question. It won't.

The default behavior of AngularJS `$http.post` performs a type check against the data property of the `$http` options. If the data is a valid JavaScript object and its content type is json, it is `JSON.stringify`-ed, otherwise the string is sent as is. ([However, this behavior does not seem to be documented.](https://docs.angularjs.org/api/ng/service/$http#post))

As of 6.7.x, Grafana implements [RxJS as its its http agent which is build on the native Fetch API](https://rxjs-dev.firebaseapp.com/api/fetch/fromFetch). Similarly, [Grafana performs a type check on the data property](https://github.com/grafana/grafana/blob/747b546c260f9a448e2cb56319f796d0301f4bb9/public/app/core/services/backend_srv.ts#L605) when a datasource request is made. Strings are sent as is.

However, Grafana performs a literal check against the headers for `application/json`. If the Content Headers contain additional params (i.e. charset), the check will fail. So instead of changing the charset params (which would be a breaking change) to satisfy this check, the data is just stringified manually.

This change should be backwards and forwards compatible.

Reply:
awesome, thanks for the explantation.